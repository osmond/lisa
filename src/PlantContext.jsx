import { createContext, useContext, useEffect, useState } from "react";
import initialPlants from "./plants.json";
import { useWeather } from "./WeatherContext.jsx";
import { getNextWateringDate } from "./utils/watering.js";
import { getWaterPlan, getSmartWaterPlan } from "./utils/waterCalculator.js";
import { cubicInchesToMl, mlToOz } from "./utils/volume.js";
import autoTag from "./utils/autoTag.js";

const PlantContext = createContext();

export const addBase = (url) => {
  if (!url) return url;
  if (/^https?:/.test(url) || url.startsWith("data:")) return url;
  const base = (process.env.VITE_BASE_PATH || "/").replace(/\/$/, "");
  if (url.startsWith(base)) return url;
  return `${base}${url.startsWith("/") ? "" : "/"}${url}`;
};

const mapPhoto = (photo) => {
  if (!photo) return photo;
  if (typeof photo === "string") {
    return { src: addBase(photo), caption: "", tags: [] };
  }
  return { ...photo, src: addBase(photo.src), tags: photo.tags || [] };
};

export function PlantProvider({ children }) {
  const [plants, setPlants] = useState(() => {
    const mapPlant = (p) => {
      const wp = p.waterPlan || { interval: 0 };
      const ml =
        wp.volume_ml !== undefined
          ? wp.volume_ml
          : cubicInchesToMl(wp.volume || 0);
      const oz =
        wp.volume_oz !== undefined ? wp.volume_oz : mlToOz(ml);
      const smart = p.smartWaterPlan
        ? {
            ...p.smartWaterPlan,
            volume_ml:
              p.smartWaterPlan.volume_ml !== undefined
                ? p.smartWaterPlan.volume_ml
                : cubicInchesToMl(p.smartWaterPlan.volume || 0),
            volume_oz:
              p.smartWaterPlan.volume_oz !== undefined
                ? p.smartWaterPlan.volume_oz
                : mlToOz(
                    p.smartWaterPlan.volume_ml !== undefined
                      ? p.smartWaterPlan.volume_ml
                      : cubicInchesToMl(p.smartWaterPlan.volume || 0),
                  ),
          }
        : null;
      return {
        ...p,
        image: addBase(p.image),
        photos: (p.photos || p.gallery || []).map(mapPhoto),
        careLog: (p.careLog || []).map((ev) => ({ ...ev, tags: ev.tags || [] })),
        diameter: p.diameter || 0,
        petSafe: !!p.petSafe,
        waterPlan: { interval: wp.interval || 0, volume_ml: Math.round(ml), volume_oz: Math.round(oz) },
        carePlan: p.carePlan || null,
        smartWaterPlan: smart,
        lastEvaluated: p.lastEvaluated || null,
      };
    };

    if (typeof localStorage !== "undefined") {
      const stored = localStorage.getItem("plants");
      if (stored) {
        try {
          return JSON.parse(stored).map(mapPlant);
        } catch {
          // fall through to initial plants
        }
      }
    }
    return initialPlants.map(mapPlant);
  });

  useEffect(() => {
    let ignore = false;
    async function load() {
      if (typeof fetch === 'undefined') return;
      try {
        const res = await fetch("/api/plants");
        if (!res.ok) return;
        const data = await res.json();
        if (!ignore) setPlants(data.map(mapPlant));
      } catch {
        // ignore network errors and keep local data
      }
    }
    load();
    return () => {
      ignore = true;
    };
  }, []);

  const weatherCtx = useWeather();
  const weather = { rainTomorrow: weatherCtx?.forecast?.rainfall || 0 };

  useEffect(() => {
    if (typeof localStorage !== "undefined") {
      localStorage.setItem("plants", JSON.stringify(plants));
    }
  }, [plants]);

  const evaluatePlans = (forecast) => {
    if (!forecast) return;
    setPlants((prev) =>
      prev.map((p) => {
        if (!p.name || !p.diameter) return p;
        const logs = (p.careLog || []).filter((l) => l.type === "Watered");
        const raw = getSmartWaterPlan(
          { name: p.name, diameter: p.diameter, light: p.light },
          forecast,
          logs,
        );
        const ml = cubicInchesToMl(raw.volume);
        const plan = {
          interval: raw.interval,
          reason: raw.reason,
          volume_ml: Math.round(ml),
          volume_oz: Math.round(mlToOz(ml)),
        };
        return {
          ...p,
          smartWaterPlan: plan,
          lastEvaluated: new Date().toISOString(),
        };
      }),
    );
  };

  useEffect(() => {
    const handler = (e) => evaluatePlans(e.detail);
    window.addEventListener("weather-updated", handler);
    return () => window.removeEventListener("weather-updated", handler);
  }, []);

  const [timelineNotes, setTimelineNotes] = useState(() => {
    if (typeof localStorage !== "undefined") {
      const stored = localStorage.getItem("timelineNotes");
      if (stored) {
        try {
          return JSON.parse(stored).map((n) => ({ ...n, tags: n.tags || [] }));
        } catch {
          // ignore
        }
      }
    }
    return [];
  });

  useEffect(() => {
    if (typeof localStorage !== "undefined") {
      localStorage.setItem("timelineNotes", JSON.stringify(timelineNotes));
    }
  }, [timelineNotes]);

  const addTimelineNote = async (text) => {
    const date = new Date().toISOString().slice(0, 10);
    const tags = await autoTag(text);
    setTimelineNotes((prev) => [...prev, { date, text, tags }]);
  };

  const logEvent = async (id, type, note = "") => {
    const date = new Date().toISOString().slice(0, 10);
    const tags = await autoTag(note);
    setPlants((prev) =>
      prev.map((p) =>
        p.id === id
          ? {
              ...p,
              careLog: [...(p.careLog || []), { date, type, note, tags }],
            }
          : p,
      ),
    );
  };

  const markWatered = async (id, note) => {
    const tags = await autoTag(note);
    const today = new Date().toISOString().slice(0, 10);
    setPlants((prev) =>
      prev.map((p) => {
        if (p.id === id) {
          const { date: nextStr, reason } = getNextWateringDate(today, weather);
          const newLog = [
            ...(p.careLog || []),
            { date: today, type: "Watered", note, tags },
          ];
          const raw = getSmartWaterPlan(
            { name: p.name, diameter: p.diameter, light: p.light },
            weatherCtx?.forecast,
            newLog.filter((l) => l.type === "Watered"),
          );
          const volMl = cubicInchesToMl(raw.volume);
          const plan = {
            interval: raw.interval,
            reason: raw.reason,
            volume_ml: Math.round(volMl),
            volume_oz: Math.round(mlToOz(volMl)),
          };
          return {
            ...p,
            careLog: newLog,
            lastWatered: today,
            nextWater: nextStr,
            nextWaterReason: reason,
            smartWaterPlan: plan,
            lastEvaluated: new Date().toISOString(),
          };
        }
        return p;
      }),
    );
  };

  const markFertilized = (id, note) => {
    setPlants((prev) =>
      prev.map((p) => {
        if (p.id === id) {
          const today = new Date().toISOString().slice(0, 10);
          let nextDate = new Date(today);
          if (p.lastFertilized && p.nextFertilize) {
            const last = new Date(p.lastFertilized);
            const next = new Date(p.nextFertilize);
            const diff = Math.round((next - last) / 86400000) || 30;
            nextDate.setDate(nextDate.getDate() + diff);
          } else {
            nextDate.setMonth(nextDate.getMonth() + 1);
          }
          return {
            ...p,
            lastFertilized: today,
            nextFertilize: nextDate.toISOString().slice(0, 10),
          };
        }
        return p;
      }),
    );
    logEvent(id, "Fertilized", note);
  };

  const addPlant = (plant) => {
    setPlants((prev) => {
      const nextId = prev.reduce((m, p) => Math.max(m, p.id), 0) + 1;

      const newPlant = {
        id: nextId,
        photos: [],
        careLog: [],
        diameter: 0,
        waterPlan: { interval: 0, volume_ml: 0, volume_oz: 0 },
        carePlan: null,
        ...plant,
      };
      if (typeof fetch !== 'undefined')
        fetch('/api/plants', {
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify(newPlant),
        }).catch(() => {})
      return [...prev, newPlant];
    });
  };

  const updatePlant = (id, updates) => {
    setPlants((prev) =>
      prev.map((p) => {
        if (p.id !== id) return p;
        const next = { ...p, ...updates };
        if (Object.prototype.hasOwnProperty.call(updates, "diameter")) {
          const base = getWaterPlan(next.name, updates.diameter, next.light);
          const ml = cubicInchesToMl(base.volume);
          next.waterPlan = {
            interval: base.interval,
            volume_ml: Math.round(ml),
            volume_oz: Math.round(mlToOz(ml)),
          };
        }
        return next;
      }),
    );
    if (typeof fetch !== 'undefined')
      fetch(`/api/plants/${id}`, {
        method: 'PUT',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(updates),
      }).catch(() => {})
  };

  const removePlant = (id) => {
    setPlants((prev) => prev.filter((p) => p.id !== id));
    if (typeof fetch !== 'undefined')
      fetch(`/api/plants/${id}`, { method: 'DELETE' }).catch(() => {})
  };

  const restorePlant = (plant, index) => {
    setPlants((prev) => {
      const arr = [...prev];
      const pos = index != null ? index : arr.length;
      arr.splice(pos, 0, plant);
      return arr;
    });
  };

  const addPhoto = async (id, photo) => {
    const newPhoto = mapPhoto(photo);
    const tags = await autoTag(newPhoto.caption || "");
    newPhoto.tags = tags;
    setPlants((prev) =>
      prev.map((p) =>
        p.id === id ? { ...p, photos: [...(p.photos || []), newPhoto] } : p,
      ),
    );
    const form = new FormData();
    if (photo.file) form.append('photos', photo.file);
    if (typeof fetch !== 'undefined' && form.has('photos'))
      fetch(`/api/plants/${id}/photos`, { method: 'POST', body: form }).catch(
        () => {},
      );
  };

  const removePhoto = (id, index) => {
    setPlants((prev) =>
      prev.map((p) => {
        if (p.id === id) {
          const updated = (p.photos || []).filter((_, i) => i !== index);
          return { ...p, photos: updated };
        }
        return p;
      }),
    );
    if (typeof fetch !== 'undefined')
      fetch(`/api/plants/${id}/photos/${index}`, { method: 'DELETE' }).catch(() => {});
  };

  return (
    <PlantContext.Provider
      value={{
        plants,
        markWatered,
        markFertilized,
        logEvent,
        addPlant,
        updatePlant,
        removePlant,
        restorePlant,
        addPhoto,
        removePhoto,
        timelineNotes,
        addTimelineNote,
      }}
    >
      {children}
    </PlantContext.Provider>
  );
}

export const usePlants = () => useContext(PlantContext);
