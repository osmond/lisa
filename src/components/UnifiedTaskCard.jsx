import React from 'react'

import { formatDaysAgo } from '../utils/dateFormat.js'

export default function UnifiedTaskCard({ plant, urgent = false, overdue = false }) {
  if (!plant) return null
  const { name, image, dueWater, dueFertilize, lastCared } = plant
  const needs = []
  if (dueWater) needs.push('water \uD83D\uDCA7')
  if (dueFertilize) needs.push('fertilizer \u2600\uFE0F')
  const summary = needs.length ? `Needs ${needs.join(' and ')}` : 'No care needed'

  const bgClass = overdue
    ? 'bg-red-50 dark:bg-red-900'
    : urgent
    ? 'bg-yellow-50 dark:bg-yellow-900'
    : 'bg-gray-50 dark:bg-gray-800'

  const lastText = lastCared ? formatDaysAgo(lastCared) : null
  const last = lastText ? (
    <p className="text-xs text-gray-500">Last cared for {lastText}</p>
  ) : null

  return (
    <div
      data-testid="unified-task-card"
      className={`rounded-2xl border border-neutral-200 dark:border-gray-600 shadow-sm overflow-hidden ${bgClass}`}
    >
      <div className="flex items-center gap-4 p-5">
        <div className="w-16 h-16 rounded-full flex items-center justify-center shadow-sm bg-neutral-100 dark:bg-gray-700">
          <img src={image} alt={name} className="w-12 h-12 rounded-full object-cover" />
        </div>
        <div className="flex-1 min-w-0">
          <p className="font-semibold font-headline text-gray-900 dark:text-gray-100 truncate">
            {name}
          </p>
          <p className="text-sm text-gray-600 dark:text-gray-300">{summary}</p>
          {last}
        </div>
        <div className="flex flex-col gap-1 items-end">
          {dueWater && (
            <button
              type="button"
              className="px-3 py-1 border border-blue-600 text-blue-600 rounded-full text-xs flex items-center gap-1"
            >
              Water Now
            </button>
          )}
          {dueFertilize && (
            <button
              type="button"
              className="px-3 py-1 border border-yellow-600 text-yellow-600 rounded-full text-xs flex items-center gap-1"
            >
              Fertilize Now
            </button>
          )}
        </div>
      </div>
    </div>
  )
}
