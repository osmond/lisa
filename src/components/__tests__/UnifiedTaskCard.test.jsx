import { render, screen, fireEvent } from '@testing-library/react'
import { MemoryRouter, useNavigate } from 'react-router-dom'
import UnifiedTaskCard from '../UnifiedTaskCard.jsx'
import { usePlants } from '../../PlantContext.jsx'

const navigateMock = jest.fn()
const markWatered = jest.fn()
const markFertilized = jest.fn()
const updatePlant = jest.fn()

jest.useFakeTimers().setSystemTime(new Date('2025-07-10'))

jest.mock('react-router-dom', () => {
  const actual = jest.requireActual('react-router-dom')
  return { ...actual, useNavigate: jest.fn() }
})

jest.mock('../../PlantContext.jsx', () => ({
  usePlants: jest.fn(),
}))

const usePlantsMock = usePlants

const plant = {
  id: 1,
  name: 'Fern',
  image: 'fern.jpg',
  lastWatered: '2025-07-10',
  dueWater: true,
  dueFertilize: false,
  lastCared: '2025-07-07',
}

beforeEach(() => {
  navigateMock.mockClear()
  markWatered.mockClear()
  markFertilized.mockClear()
  useNavigate.mockReturnValue(navigateMock)
  usePlantsMock.mockReturnValue({
    plants: [plant],
    markWatered,
    markFertilized,
    updatePlant,
  })
})

test('renders plant info and badges', () => {
  render(
    <MemoryRouter>
      <UnifiedTaskCard plant={plant} />
    </MemoryRouter>
  )
  expect(screen.getByText('Fern')).toBeInTheDocument()
  const waterBadge = screen.getByText('Water')
  expect(waterBadge).toBeInTheDocument()
  expect(waterBadge).toHaveClass('inline-flex')
  expect(screen.queryByText('Fertilize')).toBeNull()
  expect(screen.getByText('Last cared for 3 days ago')).toBeInTheDocument()
  expect(screen.queryByRole('button', { name: /mark as done/i })).toBeNull()
})

test('applies urgent style', () => {
  const { container } = render(
    <MemoryRouter>
      <UnifiedTaskCard plant={plant} urgent />
    </MemoryRouter>
  )
  const wrapper = container.querySelector('[data-testid="unified-task-card"]')
  expect(wrapper).toHaveClass('bg-sage')
})

test('applies overdue style', () => {
  const { container } = render(
    <MemoryRouter>
      <UnifiedTaskCard plant={plant} overdue />
    </MemoryRouter>
  )
  const wrapper = container.querySelector('[data-testid="unified-task-card"]')
  expect(wrapper).toHaveClass('bg-red-50')
})

test('matches snapshot in dark mode', () => {
  document.documentElement.classList.add('dark')
  const { container } = render(
    <MemoryRouter>
      <UnifiedTaskCard plant={plant} />
    </MemoryRouter>
  )
  expect(container.firstChild).toMatchSnapshot()
  document.documentElement.classList.remove('dark')
})

test('does not render a completion button', () => {
  render(
    <MemoryRouter>
      <UnifiedTaskCard plant={plant} />
    </MemoryRouter>
  )
  expect(screen.queryByRole('button', { name: /mark as done/i })).toBeNull()
})

test('partial left swipe reveals actions', () => {
  render(
    <MemoryRouter>
      <UnifiedTaskCard plant={plant} />
    </MemoryRouter>
  )
  const wrapper = screen.getByTestId('unified-task-card')
  fireEvent.pointerDown(wrapper, { clientX: 100, buttons: 1 })
  fireEvent.pointerMove(wrapper, { clientX: 70, buttons: 1 })
  expect(screen.getByRole('button', { name: /edit task/i })).toBeInTheDocument()
  fireEvent.click(screen.getByRole('button', { name: /edit task/i }))
  expect(navigateMock).toHaveBeenCalledWith('/plant/1/edit')
  fireEvent.click(screen.getByRole('button', { name: /reschedule task/i }))
  expect(updatePlant).toHaveBeenCalled()
  fireEvent.click(screen.getByRole('button', { name: /delete task/i }))
  expect(updatePlant).toHaveBeenCalledTimes(2)
})

afterAll(() => {
  jest.useRealTimers()
})
