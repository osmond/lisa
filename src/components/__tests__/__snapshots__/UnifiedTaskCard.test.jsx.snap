// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot in dark mode 1`] = `
<div
  class="relative rounded-2xl border border-neutral-200 dark:border-gray-600 shadow-sm overflow-hidden bg-gray-50 dark:bg-gray-800"
  data-testid="unified-task-card"
  style="transform: translateX(0px); transition: transform 0.2s;"
>
  <div
    class="flex items-center gap-4 p-4"
  >
    <div
      class="w-14 h-14 rounded-full flex items-center justify-center shadow-sm bg-neutral-100 dark:bg-gray-700"
    >
      <img
        alt="Fern"
        class="w-12 h-12 rounded-full object-cover"
        src="fern.jpg"
      />
    </div>
    <div
      class="flex-1 min-w-0"
    >
      <div
        class="flex items-center justify-between"
      >
        <p
          class="font-semibold font-headline text-gray-900 dark:text-gray-100 truncate"
        >
          Fern
        </p>
      </div>
      <div
        class="flex items-center gap-2 mt-1"
      >
        <span
          class="inline-flex items-center gap-1 rounded-full font-medium px-2 py-0.5 text-badge bg-water-100/90 text-water-800"
        >
          <svg
            aria-hidden="true"
            class="w-3 h-3"
            fill="currentColor"
            height="1em"
            viewBox="0 0 256 256"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <rect
              fill="none"
              height="256"
              width="256"
            />
            <path
              d="M208,144c0-72-80-128-80-128S48,72,48,144a80,80,0,0,0,160,0Z"
              fill="none"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="16"
            />
            <path
              d="M136.1,191.2a47.9,47.9,0,0,0,39.2-39.1"
              fill="none"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="16"
            />
          </svg>
          Water
        </span>
      </div>
      <p
        class="text-timestamp text-gray-500"
      >
        Last cared for 
        3 days ago
      </p>
    </div>
  </div>
  <div
    class="border-t border-neutral-200 dark:border-gray-600 px-4 py-2"
  >
    <button
      class="px-3 py-1 bg-green-600 text-white rounded-md text-sm font-medium flex items-center gap-1"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="w-4 h-4"
        fill="currentColor"
        height="1em"
        viewBox="0 0 256 256"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="none"
          height="256"
          width="256"
        />
        <polyline
          fill="none"
          points="172 104 113.3 160 84 132"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="16"
        />
        <circle
          cx="128"
          cy="128"
          fill="none"
          r="96"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="16"
        />
      </svg>
      Mark as Done
    </button>
  </div>
</div>
`;
